README.md
# Sky Color: Reflects the current sky color in the status bar

Get the latitude, longitude, and weather of the current location from [OpenWeatherMap](https://openweathermap.org/) and set the current sky color in the status bar.

![screen image](https://raw.githubusercontent.com/sumomoneko/sky-color/main/doc/sky-color.gif)

## Features

- Set the current sky color in the status bar.
- The color is whitened depending on the degree of cloudiness.

This extension is inspired by [zk-phi/sky-color-clock: Emacs clock widget](https://github.com/zk-phi/sky-color-clock).  
There are also the following projects based on the same concept:

- for Vim: [mopp/sky-color-clock.vim](https://github.com/mopp/sky-color-clock.vim)
- Desktop app: [opengl-8080/little-sky](https://github.com/opengl-8080/little-sky)

## How to setup

1. Register with [OpenWeatherMap](https://openweathermap.org/) and get an API key.
2. Set API key, Country code, and zip code in _user settings_.
   ![settings](https://raw.githubusercontent.com/sumomoneko/sky-color/main/doc/settings.png)

## Implementation Details

At startup, [Geocoding API](https://openweathermap.org/api/geocoding-api#direct_zip) is used to
obtain latitude and longitude.  
Then, every 30 minutes, [Current weather data API](https://openweathermap.org/current) is used to
obtain the time of the Sunrise/Sunset and cloud degree.  
Every 5 minutes, update the color of the status bar based on the current time.

To update colors, this extension rewrites `workbench.colorCustomizations` in User settings.json.  
The implementation for the color change is based on
[johnpapa/vscode-peacock](https://github.com/johnpapa/vscode-peacock).

## License

- MIT
- Icon Image by [Peter H](https://pixabay.com/ja/users/tama66-1032521/?utm_source=link-attribution&utm_medium=referral&utm_campaign=image&utm_content=3879030) from [Pixabay](https://pixabay.com/ja//?utm_source=link-attribution&utm_medium=referral&utm_campaign=image&utm_content=3879030)


CHANGELOG.md
# Change Log

All notable changes to the "sky-color" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [0.1.1] - 2022-12-23

- Reduce settings file updating.
- Adjust color gradation as original implementation.

## [0.1.0] - 2022-12-23

- Initial release


src/color.ts
import assert = require("assert");

// https://gist.github.com/mryhryki/29804334efd00129447878b67bff9772

export type Rgb = { r: number; g: number; b: number };
export type Hsl = { h: number; s: number; l: number };

/** get higher contrast fontcolor.
 *
 * @param bgColor background color
 * @returns BLACK or WHITE.
 */
export const getFontColor = (bgColor: Rgb): Rgb => {
  const BLACK = { r: 0, g: 0, b: 0 };
  const WHITE = { r: 1.0, g: 1.0, b: 1.0 };

  const blackRatio = getContrastRatio(bgColor, BLACK);
  const whiteRatio = getContrastRatio(bgColor, WHITE);

  return whiteRatio > blackRatio ? WHITE : BLACK;
};

/** mix color  c1 * r + s2 * (1 - r)
 *
 * @returns mixed color
 */
export const mixColor = (c1: Rgb, c2: Rgb, ratio: number): Rgb => {
  const r = c1.r * ratio + c2.r * (1.0 - ratio);
  const g = c1.g * ratio + c2.g * (1.0 - ratio);
  const b = c1.b * ratio + c2.b * (1.0 - ratio);
  return { r, g, b };
};

export const hexToRgb = (hex: string): Rgb => {
  const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
  assert(result);

  const rHex = parseInt(result[1], 16);
  const gHex = parseInt(result[2], 16);
  const bHex = parseInt(result[3], 16);

  const r = rHex / 255;
  const g = gHex / 255;
  const b = bHex / 255;
  return { r, g, b };
};

export const rgbToHex = (c: Rgb) => {
  const fmt = (c: number) => {
    return Math.round(255 * c)
      .toString(16)
      .padStart(2, "0");
  };
  return `#${fmt(c.r)}${fmt(c.g)}${fmt(c.b)}`;
};

// input: h as an angle in [0,360] and s,l in [0,1] - output: r,g,b in [0,1]
export const hslToRgb = (hsl: Hsl): Rgb => {
  const { h, s, l } = hsl;
  const a = s * Math.min(l, 1 - l);
  const f = (n: number, k = (n + h / 30) % 12) =>
    l - a * Math.max(Math.min(k - 3, 9 - k, 1), -1);
  return { r: f(0), g: f(8), b: f(4) };
};

// in: r,g,b in [0,1], out: h in [0,360) and s,l in [0,1]
export const rgbToHsl = (rgb: Rgb): Hsl => {
  const { r, g, b } = rgb;
  const v = Math.max(r, g, b);
  const c = v - Math.min(r, g, b);
  const f = 1 - Math.abs(v + v - c - 1);
  const h =
    c && (v === r ? (g - b) / c : v === g ? 2 + (b - r) / c : 4 + (r - g) / c);

  return { h: 60 * (h < 0 ? h + 6 : h), s: f ? c / f : 0, l: (v + v - c) / 2 };
};

const getRGBForCalculateLuminance = (color: number): number => {
  if (color <= 0.03928) {
    return color / 12.92;
  } else {
    return Math.pow((color + 0.055) / 1.055, 2.4);
  }
};

const getRelativeLuminance = (color: Rgb): number => {
  const { r, g, b } = color;
  const R = getRGBForCalculateLuminance(r);
  const G = getRGBForCalculateLuminance(g);
  const B = getRGBForCalculateLuminance(b);
  return 0.2126 * R + 0.7152 * G + 0.0722 * B;
};

const getContrastRatio = (color1: Rgb, color2: Rgb): number => {
  const luminance1 = getRelativeLuminance(color1);
  const luminance2 = getRelativeLuminance(color2);
  const bright = Math.max(luminance1, luminance2);
  const dark = Math.min(luminance1, luminance2);
  return (bright + 0.05) / (dark + 0.05);
};


src/weather.ts
import Ajv, { JTDDataType } from "ajv/dist/jtd";
import { Err, Ok, Result } from "./result";
import { getUnixTime, startOfDay } from "date-fns";
import { ProxyAgent, request, setGlobalDispatcher } from "undici";

export interface ApiParam {
  apiKey: string;
  zipCode: string;
  countryCode: string;
}

export type Location = { lat: number; lon: number };

export class NetworkError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "NetworkError";
  }
}
export class ApiKeyError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ApiKeyError";
  }
}
export class ParamError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ParamError";
  }
}
export class UnknownResponseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "UnknownResponseError";
  }
}

export type FetchError =
  | NetworkError
  | ApiKeyError
  | ParamError
  | UnknownResponseError;

const getSystemHttpsProxyURI = (): string | undefined => {
  return (
    process.env.HTTPS_PROXY ||
    process.env.https_proxy ||
    process.env.HTTP_PROXY ||
    process.env.http_proxy ||
    undefined
  );
};

export const setupProxy = () => {
  const httpsProxy = getSystemHttpsProxyURI();
  if (httpsProxy) {
    const proxyAgent = new ProxyAgent(httpsProxy);
    setGlobalDispatcher(proxyAgent);
  }
};

export const getLocationAsync = async (
  param: ApiParam
): Promise<Result<Location, FetchError>> => {
  const geoQueryStr = `https://api.openweathermap.org/geo/1.0/zip?zip=${encodeURIComponent(
    param.zipCode
  )},${encodeURIComponent(param.countryCode)}&appid=${encodeURIComponent(
    param.apiKey
  )}`;

  try {
    const { statusCode, body } = await request(geoQueryStr);

    switch (statusCode) {
      case 401:
        return new Err(new ApiKeyError("Invalid API key."));
      case 404:
        return new Err(new ParamError("ZIP code or country code is unknown."));
      case 200:
        break;
      default:
        return new Err(new UnknownResponseError(`status code: ${statusCode}`));
    }
    const json = await body.json();
    if (!validateLocationJson(json)) {
      return new Err(
        new UnknownResponseError(
          `location API parse error: ${validateLocationJson.errors}`
        )
      );
    }

    console.debug(`fetched: ${json}`);
    return new Ok({ lat: json.lat, lon: json.lon });
  } catch (e) {
    console.log(e);
    if (e instanceof Error) {
      return new Err(new NetworkError(e.message));
    }
    return new Err(new NetworkError("unknown error"));
  }
};

export interface Weather {
  sunrise: number;
  sunset: number;
  cloud: number;
}

export const getWeatherAsync = async (
  location: Location,
  apiKey: string
): Promise<Result<Weather, FetchError>> => {
  const weatherQueryStr = `https://api.openweathermap.org/data/2.5/weather?lat=${location.lat}&lon=${location.lon}&appid=${apiKey}`;

  try {
    const { statusCode, body } = await request(weatherQueryStr);

    switch (statusCode) {
      case 401:
        return new Err(new ApiKeyError("Invalid API key."));
      case 404:
        return new Err(new ParamError("location is invalid."));
      case 200:
        break;
      default:
        return new Err(new UnknownResponseError(`status code: ${statusCode}`));
    }
    const json = await body.json();
    if (!validateWeatherJson(json)) {
      return new Err(
        new UnknownResponseError(
          `weather API parse error: ${validateWeatherJson.errors}`
        )
      );
    }
    console.debug(`fetched: ${json}`);

    const now = new Date();
    const todayOrigin = getUnixTime(startOfDay(now));
    const sunrise = json.sys.sunrise - todayOrigin;
    const sunset = json.sys.sunset - todayOrigin;

    console.debug(
      `sunrise: ${timeFormat(sunrise)}, sunset: ${timeFormat(sunset)}, cloud: ${
        json.clouds.all
      }`
    );

    return new Ok({ sunrise, sunset, cloud: json.clouds.all });
  } catch (e) {
    if (e instanceof Error) {
      return new Err(new NetworkError(e.message));
    }
    return new Err(new NetworkError("unknown error"));
  }
};

const ajv = new Ajv();

const weatherSchema = {
  properties: {
    clouds: {
      properties: {
        all: { type: "uint8" },
      },
      additionalProperties: true,
    },
    sys: {
      properties: {
        sunrise: { type: "float64" },
        sunset: { type: "float64" },
      },
      additionalProperties: true,
    },
  },
  additionalProperties: true,
} as const;

const validateWeatherJson =
  ajv.compile<JTDDataType<typeof weatherSchema>>(weatherSchema);

const locationSchema = {
  properties: {
    lat: { type: "float64" },
    lon: { type: "float64" },
  },
  additionalProperties: true,
} as const;

const validateLocationJson =
  ajv.compile<JTDDataType<typeof locationSchema>>(locationSchema);

/** format seconds to readable string.
 *
 * @param d seconds
 * @returns formatted string ex. "12:05"
 */
const timeFormat = (d: number) => {
  const hour = Math.floor(d / 3600);
  const minute = Math.floor(d / 60) % 60;

  return `${hour.toString().padStart(2, "0")}:${minute
    .toString()
    .padStart(2, "0")}`;
};


src/extension.ts
import * as vscode from "vscode";
import { clearIntervalAsync, setIntervalAsync } from "set-interval-async";
import {
  getFontColor,
  hexToRgb,
  Hsl,
  hslToRgb,
  mixColor,
  rgbToHex,
  rgbToHsl,
} from "./color";
import {
  getLocationAsync,
  ApiParam,
  getWeatherAsync,
  Location,
  ApiKeyError,
  NetworkError,
  ParamError,
  UnknownResponseError,
  Weather,
  setupProxy,
} from "./weather";
import { getUnixTime, startOfDay } from "date-fns";

export async function activate(context: vscode.ExtensionContext) {
  const sc = new SkyColor();
  context.subscriptions.push(sc);

  context.subscriptions.push(
    vscode.workspace.onDidChangeConfiguration(() =>
      sc.onDidChangeConfiguration()
    )
  );

  await sc.onDidChangeConfiguration();
}

// eslint-disable-next-line @typescript-eslint/no-empty-function
export function deactivate() {}

class SkyColor {
  private settings_: ApiParam | undefined;
  private location_: Location | undefined;
  private weather_: Weather | undefined;

  /// periodic fetch sunrise/sunset/cloud information proc handle.
  private weatherUpdaterHandle_:
    | ReturnType<typeof setIntervalAsync>
    | undefined;

  /// periodic color updater handle.
  private colorUpdaterHandle_: ReturnType<typeof setIntervalAsync> | undefined;

  constructor() {
    setupProxy();

    this.weatherUpdaterHandle_ = setIntervalAsync(async () => {
      await this.updateWeather();
    }, 30 * 60 * 1000);

    this.colorUpdaterHandle_ = setIntervalAsync(async () => {
      await this.updateColor();
    }, 5 * 60 * 1000);
  }

  public async onDidChangeConfiguration() {
    const settings = getSettings();
    if (JSON.stringify(this.settings_) === JSON.stringify(settings)) {
      return;
    }
    this.settings_ = settings;
    await this.updateLocation();
    await this.updateWeather();
    await this.updateColor();
  }

  public dispose() {
    if (this.colorUpdaterHandle_) {
      clearIntervalAsync(this.colorUpdaterHandle_);
    }

    if (this.weatherUpdaterHandle_) {
      clearIntervalAsync(this.weatherUpdaterHandle_);
    }
  }

  private async updateLocation() {
    if (this.settings_) {
      const result = await getLocationAsync(this.settings_);
      if (result.isErr()) {
        switch (result.value.constructor) {
          case ApiKeyError:
            vscode.window.showErrorMessage(
              "Invalid/deactivated API Key. (Activation could take several hours)"
            );
            delete this.settings_;
            break;
          case NetworkError:
            // Ignore connection problems.
            break;
          case ParamError:
            vscode.window.showErrorMessage("Unknown ZIP/country code.");
            break;
          case UnknownResponseError:
            vscode.window.showErrorMessage(
              "Invalid response from openweathermap server."
            );
            break;
        }
        await this.deleteColor();
        return;
      }

      // update member
      this.location_ = { lat: result.value.lat, lon: result.value.lon };
    }
  }

  private async updateWeather() {
    if (!this.location_ || !this.settings_) {
      return;
    }
    const result = await getWeatherAsync(this.location_, this.settings_.apiKey);
    if (result.isErr()) {
      switch (result.value.constructor) {
        case ApiKeyError:
          delete this.settings_;
          break;
        case NetworkError:
        case ParamError:
        case UnknownResponseError:
          console.warn(result.value.message);
          break;
      }
      return;
    }
    this.weather_ = result.value;
  }

  private async deleteColor() {
    const inspect = vscode.workspace
      .getConfiguration()
      .inspect("workbench.colorCustomizations");

    if (!inspect) {
      return;
    }

    const currentValues: Partial<{ [key: string]: unknown }> =
      inspect.globalValue ?? {};

    delete currentValues["statusBar.background"];
    delete currentValues["statusBar.foreground"];

    const workspaceConfig = vscode.workspace.getConfiguration();
    await workspaceConfig.update(
      "workbench.colorCustomizations",
      currentValues,
      vscode.ConfigurationTarget.Global
    );
  }

  private async updateColor() {
    if (!this.weather_) {
      return;
    }
    const now = new Date();
    const currentTime = getUnixTime(now) - getUnixTime(startOfDay(now));

    const background = getSkyColor(currentTime, this.weather_);
    const foreground = getFontColor(background);

    const newColorValues = {
      // eslint-disable-next-line @typescript-eslint/naming-convention
      "statusBar.background": rgbToHex(background),
      // eslint-disable-next-line @typescript-eslint/naming-convention
      "statusBar.foreground": rgbToHex(foreground),
    };

    const inspect = vscode.workspace
      .getConfiguration()
      .inspect("workbench.colorCustomizations");

    if (!inspect) {
      return {};
    }

    const currentValues: object = inspect.globalValue ?? {};
    const updatedValues = Object.assign({}, currentValues);
    Object.assign(updatedValues, newColorValues);

    if (JSON.stringify(currentValues) === JSON.stringify(updatedValues)) {
      console.debug("no rewrite required.");
      return;
    }

    const workspaceConfig = vscode.workspace.getConfiguration();
    await workspaceConfig.update(
      "workbench.colorCustomizations",
      updatedValues,
      vscode.ConfigurationTarget.Global
    );
    console.debug("config updated.");
  }
}

const getSettings = (): ApiParam | undefined => {
  const config = vscode.workspace.getConfiguration("sky-color");
  const zipCode: string = config.get("locationZipCode") ?? "1000000";
  const countryCode: string =
    vscode.workspace.getConfiguration("sky-color").get("locationCountryCode") ??
    "JP";
  const apiKey: string | undefined = vscode.workspace
    .getConfiguration("sky-color")
    .get("apiKey");

  if (!apiKey) {
    return;
  }

  return { zipCode, countryCode, apiKey };
};

const getSkyColor = (currentTime: number, weather: Weather) => {
  const { sunrise, sunset, cloud } = weather;
  const keyColors = [
    { offset: 0, color: "#111111" },
    { offset: sunrise - 2.0 * 60 * 60, color: "#111111" },
    { offset: sunrise - 1.5 * 60 * 60, color: "#4d548a" },
    { offset: sunrise - 1.0 * 60 * 60, color: "#c486b1" },
    { offset: sunrise - 0.5 * 60 * 60, color: "#ee88a0" },
    { offset: sunrise, color: "#ff7d75" },
    { offset: sunrise + 0.5 * 60 * 60, color: "#f4eeef" },
    { offset: (sunrise + sunset) / 2, color: "#8bd7f5" },
    { offset: sunset - 1.2 * 60 * 60, color: "#b3f2e6" },
    { offset: sunset - 0.8 * 60 * 60, color: "#f3e693" },
    { offset: sunset - 0.4 * 60 * 60, color: "#f86b10" },
    { offset: sunset, color: "#100028" },
    { offset: sunset + 0.5 * 60 * 60, color: "#111111" },
    { offset: 24 * 60 * 60, color: "#111111" },
  ];

  let i = keyColors.findIndex((elem) => elem.offset > currentTime);
  if (i <= 0) {
    i = 1;
  }

  const ratio =
    (currentTime - keyColors[i - 1].offset) /
    (keyColors[i].offset - keyColors[i - 1].offset);
  const clearSkyColor = mixColor(
    hexToRgb(keyColors[i - 1].color),
    hexToRgb(keyColors[i].color),
    1.0 - ratio
  );

  const skyHsl = rgbToHsl(clearSkyColor);
  const cloudedHsl: Hsl = {
    h: skyHsl.h,
    s: skyHsl.s - ((skyHsl.s * cloud) / 100) * 0.9,
    l: Math.min(0.95, (cloud / 100) * 0.15 + skyHsl.l),
  };
  const cloudedColor = hslToRgb(cloudedHsl);

  return cloudedColor;
};
